Upstream WebRTC AV1 decoder support

youennf opened the pull request:
#### 9886ad7abe71b691066baa91b8ae6e13a1014063
<pre>
Upstream WebRTC AV1 decoder support
<a href="https://bugs.webkit.org/show_bug.cgi?id=262481">https://bugs.webkit.org/show_bug.cgi?id=262481</a>
<a href="https://rdar.apple.com/116343842">rdar://116343842</a>

Reviewed by Eric Carlson.

Add Source/WebCore/platform/video-codecs/cocoa folder for WebRTC/WebCodecs platform decoders and encoders
Add AV1 VTB decoder support class as RTCVideoDecoderVTBAV1.
In the future, we should add H264 and H265 encoders and decoders as well.

* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.mm:
(WebCore::av1HardwareDecoderAvailable):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::hasVP9ExtensionSupport): Deleted.
* Source/WebCore/platform/video-codecs/BitReader.cpp: Copied from Source/WebCore/platform/mediastream/cocoa/WebRTCVideoDecoder.h.
(WebCore::BitReader::read):
(WebCore::BitReader::readBit):
* Source/WebCore/platform/video-codecs/BitReader.h: Copied from Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h.
(WebCore::BitReader::BitReader):
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.h: Copied from Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h.
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm: Added.
(BitReader::BitReader):
(BitReader::read):
(BitReader::readBit):
(readULEBSize):
(getSequenceHeaderOBU):
(parseSequenceHeaderOBU):
(computeAV1InputFormat):
(av1BufferToCMSampleBuffer):
(av1DecompressionOutputCallback):
(-[RTCVideoDecoderVTBAV1 init]):
(-[RTCVideoDecoderVTBAV1 dealloc]):
(-[RTCVideoDecoderVTBAV1 setWidth:height:]):
(-[RTCVideoDecoderVTBAV1 decodeData:size:timeStamp:]):
(-[RTCVideoDecoderVTBAV1 setCallback:]):
(-[RTCVideoDecoderVTBAV1 setError:]):
(-[RTCVideoDecoderVTBAV1 releaseDecoder]):
(-[RTCVideoDecoderVTBAV1 resetDecompressionSession]):
(-[RTCVideoDecoderVTBAV1 configureDecompressionSession]):
(-[RTCVideoDecoderVTBAV1 destroyDecompressionSession]):
(-[RTCVideoDecoderVTBAV1 flush]):
* Source/WebCore/platform/video-codecs/cocoa/WebRTCVideoDecoder.h: Renamed from Source/WebCore/platform/mediastream/cocoa/WebRTCVideoDecoder.h.
* Source/WebCore/platform/video-codecs/cocoa/WebRTCVideoDecoder.mm: Renamed from Source/WebCore/platform/mediastream/cocoa/WebRTCVideoDecoder.mm.
(WebCore::createAV1VTBDecoder):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_preferences):
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::createLocalDecoder):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::createGPUProcessConnection):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didInitialize):
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.serialization.in:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::createVideoDecoder):
(WebKit::LibWebRTCCodecs::videoCodecTypeFromWebCodec):
(WebKit::LibWebRTCCodecs::setHasAV1HardwareDecoder):
(WebKit::LibWebRTCCodecs::setHasVP9ExtensionSupport): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
(WebKit::LibWebRTCCodecs::hasAV1HardwareDecoder const):
(WebKit::LibWebRTCCodecs::hasVP9ExtensionSupport const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):

Canonical link: <a href="https://commits.webkit.org/274002@main">https://commits.webkit.org/274002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4280116f5c83a2a11fa06d8217449f13511ef812

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37461 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39724 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39987 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33383 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13504 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31791 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38026 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32875 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11990 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11977 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33554 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41250 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33866 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33974 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37869 "Found 2 new test failures: http/tests/inspector/network/har/har-page.html, http/wpt/mediastream/mediastreamtrackprocessor-videoframe-timestamp.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12533 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10110 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36025 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/14000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8448 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12944 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.h:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
```
2023?

youennf replied:
I guess we could do 2022-2023.
I kept 2022 here.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
```diff
@@ -807,11 +807,13 @@ static uint8_t convertSubsamplingXYToChromaSubsampling(uint64_t x, uint64_t y)
 }
 
-bool hasVP9ExtensionSupport()
+bool av1HardwareDecoderAvailable()
```
this would be better in an AV1UtilitiesCocoa.

There's VP9 and HEVC already

youennf replied:
Done

jyavenard commented on Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
```diff
@@ -0,0 +1,489 @@
+typedef struct OpaqueVTDecompressionSession*  VTDecompressionSessionRef;
+typedef void (*VTDecompressionOutputCallback)(void * decompressionOutputRefCon, void * sourceFrameRefCon, OSStatus status, VTDecodeInfoFlags infoFlags, CVImageBufferRef imageBuffer, CMTime presentationTimeStamp, CMTime presentationDuration);
+
+class BitReader {
```
this should be in a file on its own I think, it's not the first time I saw a BitReader

youennf replied:
Done

jyavenard commented on Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
```diff
@@ -0,0 +1,489 @@
+
+std::optional<uint64_t> BitReader::read(size_t bits)
+{
+    // FIXME: We should optimize this routine.
```
should assert that this is never greater than 64

youennf replied:
Done

jyavenard commented on Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
```diff
@@ -70,45 +66,59 @@ static webrtc::WebKitVideoDecoder createVideoDecoder(const webrtc::SdpVideoForma
 
 std::optional<VideoCodecType> LibWebRTCCodecs::videoCodecTypeFromWebCodec(const String& codec)
 {
-    // WebCodecs is assuming case sensitive comparisons.
```
is this no longer the case?

youennf replied:
I readded it.

jyavenard commented on Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
```diff
@@ -70,45 +66,59 @@ static webrtc::WebKitVideoDecoder createVideoDecoder(const webrtc::SdpVideoForma
+    m_hasAV1HardwareDecoder = hasAV1HardwareDecoder;
+    if (!hasAV1HardwareDecoder)
+        return;
+    Page::forEachPage([](auto& page) {
```
shouldn't this be using a PlatformStrategy instead (a singleton)?

youennf replied:
We can rework it in a follow-up if we think this is useful.

youennf committed:
Upstream WebRTC AV1 decoder support
https://bugs.webkit.org/show_bug.cgi?id=262481
rdar://116343842

Reviewed by Eric Carlson.

Add Source/WebCore/platform/video-codecs/cocoa folder for WebRTC/WebCodecs platform decoders and encoders
Add AV1 VTB decoder support class as RTCVideoDecoderVTBAV1.
In the future, we should add H264 and H265 encoders and decoders as well.

* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.mm:
(WebCore::av1HardwareDecoderAvailable):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::hasVP9ExtensionSupport): Deleted.
* Source/WebCore/platform/video-codecs/BitReader.cpp: Copied from Source/WebCore/platform/mediastream/cocoa/WebRTCVideoDecoder.h.
(WebCore::BitReader::read):
(WebCore::BitReader::readBit):
* Source/WebCore/platform/video-codecs/BitReader.h: Copied from Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h.
(WebCore::BitReader::BitReader):
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.h: Copied from Source/WebCore/platform/graphics/cocoa/AV1UtilitiesCocoa.h.
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm: Added.
(BitReader::BitReader):
(BitReader::read):
(BitReader::readBit):
(readULEBSize):
(getSequenceHeaderOBU):
(parseSequenceHeaderOBU):
(computeAV1InputFormat):
(av1BufferToCMSampleBuffer):
(av1DecompressionOutputCallback):
(-[RTCVideoDecoderVTBAV1 init]):
(-[RTCVideoDecoderVTBAV1 dealloc]):
(-[RTCVideoDecoderVTBAV1 setWidth:height:]):
(-[RTCVideoDecoderVTBAV1 decodeData:size:timeStamp:]):
(-[RTCVideoDecoderVTBAV1 setCallback:]):
(-[RTCVideoDecoderVTBAV1 setError:]):
(-[RTCVideoDecoderVTBAV1 releaseDecoder]):
(-[RTCVideoDecoderVTBAV1 resetDecompressionSession]):
(-[RTCVideoDecoderVTBAV1 configureDecompressionSession]):
(-[RTCVideoDecoderVTBAV1 destroyDecompressionSession]):
(-[RTCVideoDecoderVTBAV1 flush]):
* Source/WebCore/platform/video-codecs/cocoa/WebRTCVideoDecoder.h: Renamed from Source/WebCore/platform/mediastream/cocoa/WebRTCVideoDecoder.h.
* Source/WebCore/platform/video-codecs/cocoa/WebRTCVideoDecoder.mm: Renamed from Source/WebCore/platform/mediastream/cocoa/WebRTCVideoDecoder.mm.
(WebCore::createAV1VTBDecoder):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_preferences):
* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::createLocalDecoder):
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::createGPUProcessConnection):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::didInitialize):
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnectionInfo.serialization.in:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::createVideoDecoder):
(WebKit::LibWebRTCCodecs::videoCodecTypeFromWebCodec):
(WebKit::LibWebRTCCodecs::setHasAV1HardwareDecoder):
(WebKit::LibWebRTCCodecs::setHasVP9ExtensionSupport): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
(WebKit::LibWebRTCCodecs::hasAV1HardwareDecoder const):
(WebKit::LibWebRTCCodecs::hasVP9ExtensionSupport const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):

Canonical link: https://commits.webkit.org/274002@main

cdumez commented on Source/WebCore/platform/video-codecs/BitReader.h:
```diff
@@ -0,0 +1,48 @@
+ */
+
+#pragma once
+
```
Please don't forget to include what you use (in this case optional and span). I had to fix in a follow-up.
